internal/sink/sink.go

```diff
@@ -0,0 +1,544 @@
+package sink
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/signal"
+	"regexp"
+	"syscall"
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
+
+	"github.com/openmeterio/openmeter/internal/dedupe"
+	"github.com/openmeterio/openmeter/internal/ingest/kafkaingest/serializer"
+)
+
+var namespaceTopicRegexp = regexp.MustCompile(`^om_([A-Za-z0-9]+(?:_[A-Za-z0-9]+)*)_events$`)
+var defaultDeadletterTopicTemplate = "om_%s_events_deadletter"
+
+type SinkMessage struct {
+	Namespace    string
+	KafkaMessage *kafka.Message
+	Serialized   *serializer.CloudEventsKafkaPayload
+	Error        *ProcessingError
+}
+
+type Sink struct {
+	consumer         *kafka.Consumer
+	producer         *kafka.Producer
+	config           *SinkConfig
+	running          bool
+	buffer           SinkBuffer
+	flushTimer       *time.Timer
+	namespaceStore   *NamespaceStore
+	namespaceRefetch *time.Timer
+}
+
+type SinkConfig struct {
+	Context                 context.Context
+	Logger                  *slog.Logger
+	Storage                 Storage
+	Deduplicator            dedupe.Deduplicator
+	ConsumerKafkaConfig     kafka.ConfigMap
+	ProducerKafkaConfig     kafka.ConfigMap
+	MinCommitCount          int
+	MaxCommitWait           time.Duration
+	NamespaceRefetch        time.Duration
+	DeadletterTopicTemplate string
+}
+
+func NewSink(config *SinkConfig) (*Sink, error) {
+	if config.Deduplicator == nil {
+		config.Logger.Warn("deduplicator is not set, deduplication will be disabled")
+	}
+
+	// These are Kafka configs but also related to sink logic
+	_ = config.ConsumerKafkaConfig.SetKey("session.timeout.ms", 6000)
+	_ = config.ConsumerKafkaConfig.SetKey("enable.auto.commit", false)
+	_ = config.ConsumerKafkaConfig.SetKey("enable.auto.offset.store", false)
+	_ = config.ConsumerKafkaConfig.SetKey("go.application.rebalance.enable", true)
+
+	consumer, err := kafka.NewConsumer(&config.ConsumerKafkaConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create consumer: %s", err)
+	}
+
+	producer, err := kafka.NewProducer(&config.ProducerKafkaConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create producer: %w", err)
+	}
```

Comment:
It would be nicer to inject these.

Reply:
Also, do we use the producer for anything, but deadletter? Would it make sense to introduce an interface for invalid messages and make Kafka an implementation detail there? We could add other reporting options in the future?

Reply:
Good idea, created an issue for future improvements: https://github.com/openmeterio/openmeter/issues/322